# Sprint 3 – Advanced Statistics & Feature Engineering


## Sprint Goal

### Validate analytical findings using statistical techniques and engineer meaningful features that improve 
### operational analysis and flight performance evaluation.


## Tasks

# 1. Statistical Analysis

### Perform suitable statistical analysis on the dataset.
#### Include:


# ➤Measures of Central Tendency

### Load the cleaned dataset

In [1]:
import pandas as pd
import numpy as np

df = pd.read_csv("cleaned_flight_dataset.csv")

print("Dataset Shape:", df.shape)

Dataset Shape: (50000, 31)


In [2]:
df.head()

,YEAR,MONTH,DAY,DAY_OF_WEEK,AIRLINE,FLIGHT_NUMBER,TAIL_NUMBER,ORIGIN_AIRPORT,DESTINATION_AIRPORT,SCHEDULED_DEPARTURE,...,ARRIVAL_TIME,ARRIVAL_DELAY,DIVERTED,CANCELLED,CANCELLATION_REASON,AIR_SYSTEM_DELAY,SECURITY_DELAY,AIRLINE_DELAY,LATE_AIRCRAFT_DELAY,WEATHER_DELAY
0,2015,5,14,4,US,1721,N174US,CLT,SMF,1810,...,2043.0,4.0,0,0,Not Cancelled,0.0,0.0,0.0,0.0,0.0
1,2015,11,29,7,MQ,3363,N902MQ,LEX,ORD,1837,...,1942.0,17.0,0,0,Not Cancelled,0.0,0.0,0.0,17.0,0.0
2,2015,2,10,2,US,1933,N562UW,PHL,ORD,730,...,834.0,-28.0,0,0,Not Cancelled,0.0,0.0,0.0,0.0,0.0
3,2015,12,21,1,UA,1929,N460UA,SEA,EWR,622,...,1414.0,-17.0,0,0,Not Cancelled,0.0,0.0,0.0,0.0,0.0
4,2015,5,7,4,DL,1478,N67171,ATL,PHL,725,...,920.0,-5.0,0,0,Not Cancelled,0.0,0.0,0.0,0.0,0.0


### Select important numerical variables

In [3]:
central_cols = [
    'DEPARTURE_DELAY',
    'ARRIVAL_DELAY',
    'DISTANCE',
    'AIR_TIME',
    'TAXI_OUT',
    'TAXI_IN'
]

### Calculate Mean, Median and Mode

In [4]:
mean_values = df[central_cols].mean()
median_values = df[central_cols].median()
mode_values = df[central_cols].mode().iloc[0]

### Create a summary table

In [5]:
central_tendency = pd.DataFrame({
    'Mean': mean_values,
    'Median': median_values,
    'Mode': mode_values
})

central_tendency.round(2)

,Mean,Median,Mode
DEPARTURE_DELAY,9.09,-2.0,-3.0
ARRIVAL_DELAY,4.00,-5.0,-10.0
DISTANCE,824.20,646.0,337.0
AIR_TIME,113.81,94.0,63.0
TAXI_OUT,16.02,14.0,11.0
TAXI_IN,7.41,6.0,4.0


#### 1- Departure delay has a mean of 9.09 minutes, while the median is -2 minutes, indicating that extreme delays increase the average.
#### 2- Arrival delay has a mean of 4.00 minutes compared with a median of -5 minutes, showing that a smaller number of large delays affect
#### the average.
#### 3- Flight distance has a mean of 824.20 miles and a median of 646 miles, indicating that longer flights influence the average distance.
#### 4- Air time has a mean of 113.81 minutes compared with a median of 94 minutes, suggesting variation in flight durations.
#### 5- Taxi-out and taxi-in times have relatively close mean and median values, indicating less influence from extreme observations.
#### 6- Overall, the differences between mean and median are more noticeable for delay-related variables, suggesting skewed delay 
#### distributions.

## ➤Measures of Dispersion


### Calculate Range

In [6]:
range_values = df[central_cols].max() - df[central_cols].min()

range_values.round(2)

DEPARTURE_DELAY    1570.0
ARRIVAL_DELAY      1644.0
DISTANCE           4952.0
AIR_TIME            645.0
TAXI_OUT            148.0
TAXI_IN             164.0
dtype: float64

#### 1- Arrival delay has a wide range.
#### 2- Departure delay also shows high variation.
#### 3- Distance has the largest range due to different flight lengths.
#### 4- Air time varies considerably across flights.
#### 5- Taxi-out and taxi-in have comparatively smaller ranges.
#### 6- Overall, the range shows considerable variation in flight operations.

### Calculate Variance

In [7]:
variance_values = df[central_cols].var()

variance_values.round(2)

DEPARTURE_DELAY      1257.75
ARRIVAL_DELAY        1425.44
DISTANCE           374407.33
AIR_TIME             5264.86
TAXI_OUT               77.54
TAXI_IN                31.49
dtype: float64

#### 1- Arrival delay shows high variability.
#### 2- Departure delay also has considerable variation.
#### 3- Distance has the highest variance due to different flight lengths.
#### 4- Air time varies across different flights.
#### 5- Taxi-out and taxi-in have comparatively lower variance.
#### 6- Overall, flight operations show considerable variation.

### Calculate Standard Deviation

In [8]:
std_values = df[central_cols].std()

std_values.round(2)

DEPARTURE_DELAY     35.46
ARRIVAL_DELAY       37.75
DISTANCE           611.89
AIR_TIME            72.56
TAXI_OUT             8.81
TAXI_IN              5.61
dtype: float64

#### 1- Arrival delay shows high variability.
#### 2- Departure delay also has considerable spread.
#### 3- Distance has the highest standard deviation due to different flight lengths.
#### 4- Air time varies considerably across flights.
#### 5- Taxi-out and taxi-in have comparatively lower variability.
#### 6- Overall, flight operations show noticeable variation.

### Calculate IQR

In [9]:
iqr_values = (
    df[central_cols].quantile(0.75)
    - df[central_cols].quantile(0.25)
)

iqr_values.round(2)

DEPARTURE_DELAY     12.0
ARRIVAL_DELAY       21.0
DISTANCE           693.0
AIR_TIME            83.0
TAXI_OUT             8.0
TAXI_IN              5.0
dtype: float64

#### 1- Arrival delay has an IQR of 21 minutes.
#### 2- Departure delay has an IQR of 12 minutes.
#### 3- Distance has an IQR of 693 miles.
#### 4- Air time has an IQR of 83 minutes.
#### 5- Taxi-out and taxi-in have smaller IQR values.
#### 6- IQR shows the spread of the middle 50% of flights.

### Create one summary table

In [10]:
dispersion = pd.DataFrame({
    'Range': range_values,
    'Variance': variance_values,
    'Standard Deviation': std_values,
    'IQR': iqr_values
})

dispersion.round(2)

,Range,Variance,Standard Deviation,IQR
DEPARTURE_DELAY,1570.0,1257.75,35.46,12.0
ARRIVAL_DELAY,1644.0,1425.44,37.75,21.0
DISTANCE,4952.0,374407.33,611.89,693.0
AIR_TIME,645.0,5264.86,72.56,83.0
TAXI_OUT,148.0,77.54,8.81,8.0
TAXI_IN,164.0,31.49,5.61,5.0


#### 1- Arrival and departure delays show considerable variation across flights.
#### 2- Arrival delay has slightly higher spread than departure delay.
#### 3- Distance has the highest variability because flight lengths differ widely.
#### 4- Air time also shows noticeable variation across flights.
#### 5- Taxi-out and taxi-in times have comparatively lower variability.
#### 6- Overall, delay and flight-operation variables show different levels of spread.

## ➤Distribution Analysis


### Check Skewness

In [11]:
skewness = df[central_cols].skew()

skewness.round(2)

DEPARTURE_DELAY    7.27
ARRIVAL_DELAY      6.39
DISTANCE           1.45
AIR_TIME           1.41
TAXI_OUT           3.39
TAXI_IN            5.42
dtype: float64

#### 1- All selected variables show positive skewness.
#### 2- Departure delay has strong right skewness.
#### 3- Arrival delay also has strong right skewness.
#### 4- Taxi-out and taxi-in show noticeable right skewness.
#### 5- Distance and air time have comparatively lower skewness.
#### 6- Overall, most variables contain higher values that extend the right tail.

### Check Kurtosis

In [12]:
kurtosis = df[central_cols].kurtosis()

kurtosis.round(2)

DEPARTURE_DELAY    125.86
ARRIVAL_DELAY      109.38
DISTANCE             2.48
AIR_TIME             2.37
TAXI_OUT            23.00
TAXI_IN             67.91
dtype: float64

#### 1- Departure delay has very high kurtosis, showing extreme values.
#### 2- Arrival delay also has very high kurtosis due to severe delays.
#### 3- Taxi-in and taxi-out show strong presence of extreme values.
#### 5- Distance and air time have comparatively lower kurtosis.
#### 6- Higher kurtosis indicates heavier tails and more extreme observations.
#### 7- Overall, several flight-operation variables contain significant outliers.

### Create a summary table

In [13]:
distribution_analysis = pd.DataFrame({
    'Skewness': skewness,
    'Kurtosis': kurtosis
})

distribution_analysis.round(2)

,Skewness,Kurtosis
DEPARTURE_DELAY,7.27,125.86
ARRIVAL_DELAY,6.39,109.38
DISTANCE,1.45,2.48
AIR_TIME,1.41,2.37
TAXI_OUT,3.39,23.00
TAXI_IN,5.42,67.91


#### 1- All variables show positive skewness.
#### 2- Departure and arrival delays are highly right-skewed.
#### 3- Departure and arrival delays also have very high kurtosis.
#### 4- Taxi-out and taxi-in contain noticeable extreme values.
#### 5- Distance and air time show comparatively lower skewness and kurtosis.
#### 6- Overall, several operational variables contain extreme observations.

## ➤Correlation

### Calculate correlation matrix

In [14]:
correlation_matrix = df[central_cols].corr()

correlation_matrix.round(2)

,DEPARTURE_DELAY,ARRIVAL_DELAY,DISTANCE,AIR_TIME,TAXI_OUT,TAXI_IN
DEPARTURE_DELAY,1.00,0.94,0.02,0.03,0.07,0.00
ARRIVAL_DELAY,0.94,1.00,-0.03,-0.01,0.24,0.11
DISTANCE,0.02,-0.03,1.00,0.99,0.08,0.08
AIR_TIME,0.03,-0.01,0.99,1.00,0.09,0.09
TAXI_OUT,0.07,0.24,0.08,0.09,1.00,-0.00
TAXI_IN,0.00,0.11,0.08,0.09,-0.00,1.00


#### 1- Departure delay is strongly related to arrival delay.
#### 2- Distance and air time have a strong positive relationship.
#### 3- Taxi-out has a weak positive relationship with arrival delay.
#### 4- Taxi-in has a weak positive relationship with arrival delay.
#### 5- Distance has almost no relationship with arrival delay.
#### 6- Overall, departure delay has the strongest relationship with arrival delay.

### Focus on Arrival Delay

In [15]:
arrival_delay_corr = (
    correlation_matrix['ARRIVAL_DELAY']
    .sort_values(ascending=False)
)

arrival_delay_corr.round(2)

ARRIVAL_DELAY      1.00
DEPARTURE_DELAY    0.94
TAXI_OUT           0.24
TAXI_IN            0.11
AIR_TIME          -0.01
DISTANCE          -0.03
Name: ARRIVAL_DELAY, dtype: float64

### Visualize the correlation matrix

In [16]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

#### 1- Departure delay has the strongest positive correlation with arrival delay (0.94).
#### 2- Taxi-out time has a weak positive correlation with arrival delay (0.24).
#### 3- Taxi-in time has a weak positive correlation with arrival delay (0.11).
#### 4- Air time has almost no correlation with arrival delay (-0.01).
#### 5- Distance has almost no correlation with arrival delay (-0.03).
#### 6- Overall, departure delay is the main variable associated with arrival delay.

plt.figure(figsize=(10, 6))

sns.heatmap(
    correlation_matrix,
    annot=True,
    cmap='coolwarm',
    fmt='.2f'
)

plt.title("Correlation Matrix of Flight Operational Variables")
plt.tight_layout()
plt.show()

## ➤Covariance

In [17]:
covariance_matrix = df[central_cols].cov()

covariance_matrix.round(2)

,DEPARTURE_DELAY,ARRIVAL_DELAY,DISTANCE,AIR_TIME,TAXI_OUT,TAXI_IN
DEPARTURE_DELAY,1257.75,1251.86,536.60,64.33,21.07,0.21
ARRIVAL_DELAY,1251.86,1425.44,-647.39,-22.43,78.56,22.65
DISTANCE,536.60,-647.39,374407.33,43833.78,409.40,269.35
AIR_TIME,64.33,-22.43,43833.78,5264.86,57.21,34.70
TAXI_OUT,21.07,78.56,409.40,57.21,77.54,-0.06
TAXI_IN,0.21,22.65,269.35,34.70,-0.06,31.49


#### 1- Departure delay and arrival delay have strong positive covariance.
#### 2- Distance and air time have strong positive covariance.
#### 3- Arrival delay and taxi-out time have positive covariance.
#### 4- Arrival delay and distance have negative covariance.
#### 5- Air time and arrival delay show slightly negative covariance.
#### 6- Overall, covariance shows how operational variables change together.

#  2. Hypothesis Testing


### Perform appropriate hypothesis tests wherever applicable.
#### Examples:


## ➤Z-Test


### Define the hypotheses

#### 1- Null Hypothesis (H₀):The average arrival delay is 0 minutes.
#### 2- Alternate Hypothesis (H₁):The average arrival delay is different from 0 minutes.
#### 3- Significance Level (α):0.05

### Perform the Z-Test

In [18]:
from statsmodels.stats.weightstats import ztest

arrival_delay = df['ARRIVAL_DELAY'].dropna()

z_stat, p_value = ztest(
    arrival_delay,
    value=0,
    alternative='two-sided'
)

print("Z-Statistic:", round(z_stat, 4))
print("P-Value:", round(p_value, 6))

Z-Statistic: 23.4693
P-Value: 0.0


### Decision

In [19]:
alpha = 0.05

if p_value < alpha:
    print("Reject the Null Hypothesis")
else:
    print("Fail to Reject the Null Hypothesis")

Reject the Null Hypothesis


#### 1- Since the p-value (0.0) is less than the significance level (0.05), we reject the Null Hypothesis.
#### 2- There is statistically significant evidence that the average arrival delay is different from 0 minutes.

### Business Interpretation

#### 1- The average arrival delay is statistically significant.
#### 2- Flights are not consistently arriving exactly on schedule.
#### 3- Airlines should focus on reducing operational delays.
#### 4- Improving departure and turnaround processes can help improve arrival punctuality.

## ➤T-Test

### Hypotheses

#### 1- Null Hypothesis (H₀):The average arrival delay is 0 minutes.
#### 2- Alternate Hypothesis (H₁):The average arrival delay is different from 0 minutes.
#### 3- Significance Level (α):0.05

### Perform the T-Test

In [20]:
from scipy.stats import ttest_1samp

arrival_delay = df['ARRIVAL_DELAY'].dropna()

t_stat, p_value = ttest_1samp(
    arrival_delay,
    popmean=0
)

print("T-Statistic:", round(t_stat, 4))
print("P-Value:", round(p_value, 6))

T-Statistic: 23.4693
P-Value: 0.0


### Decision

In [21]:
alpha = 0.05

if p_value < alpha:
    print("Reject the Null Hypothesis")
else:
    print("Fail to Reject the Null Hypothesis")

Reject the Null Hypothesis


#### 1- The p-value (0.0) is less than 0.05.
#### 2- Therefore, we reject the Null Hypothesis.
#### 3- There is statistically significant evidence that the average arrival delay differs from 0 minutes.

### Business Interpretation

#### 1- Average arrival delay is statistically significant.
#### 2- Flight punctuality differs from the on-time target.
#### 3- Airlines should focus on reducing operational delays.
#### 4- Improving departure and turnaround processes can improve arrival performance.

## ➤Chi-Square Test

### Create the contingency table

In [22]:
print(df.columns.tolist())

['YEAR', 'MONTH', 'DAY', 'DAY_OF_WEEK', 'AIRLINE', 'FLIGHT_NUMBER', 'TAIL_NUMBER', 'ORIGIN_AIRPORT', 'DESTINATION_AIRPORT', 'SCHEDULED_DEPARTURE', 'DEPARTURE_TIME', 'DEPARTURE_DELAY', 'TAXI_OUT', 'WHEELS_OFF', 'SCHEDULED_TIME', 'ELAPSED_TIME', 'AIR_TIME', 'DISTANCE', 'WHEELS_ON', 'TAXI_IN', 'SCHEDULED_ARRIVAL', 'ARRIVAL_TIME', 'ARRIVAL_DELAY', 'DIVERTED', 'CANCELLED', 'CANCELLATION_REASON', 'AIR_SYSTEM_DELAY', 'SECURITY_DELAY', 'AIRLINE_DELAY', 'LATE_AIRCRAFT_DELAY', 'WEATHER_DELAY']


In [23]:
def delay_category(delay):
    if delay <= 0:
        return 'On Time/Early'
    elif delay <= 15:
        return 'Minor Delay'
    elif delay <= 60:
        return 'Moderate Delay'
    else:
        return 'Severe Delay'

df['DELAY_CATEGORY'] = df['ARRIVAL_DELAY'].apply(delay_category)

In [24]:
df['DELAY_CATEGORY'].value_counts()

DELAY_CATEGORY
On Time/Early     31227
Minor Delay        9083
Moderate Delay     6165
Severe Delay       3525
Name: count, dtype: int64

In [25]:
from scipy.stats import chi2_contingency

crosstab = pd.crosstab(
    df['AIRLINE'],
    df['DELAY_CATEGORY']
)

crosstab

DELAY_CATEGORY,Minor Delay,Moderate Delay,On Time/Early,Severe Delay
AIRLINE,,,,
AA,1115,731,3950,412
AS,279,159,1029,48
B6,395,310,1410,204
DL,1207,665,5256,360
EV,926,644,2939,460
F9,133,123,427,67
HA,200,65,404,9
MQ,370,351,1515,323
NK,224,208,451,107


In [26]:
chi2, p_value, dof, expected = chi2_contingency(crosstab)

print("Chi-Square Statistic:", round(chi2, 4))
print("P-Value:", round(p_value, 6))
print("Degrees of Freedom:", dof)

Chi-Square Statistic: 785.4879
P-Value: 0.0
Degrees of Freedom: 39


In [27]:
alpha = 0.05

if p_value < alpha:
    print("Reject the Null Hypothesis")
else:
    print("Fail to Reject the Null Hypothesis")

Reject the Null Hypothesis


#### 1- The p-value (0.0) is less than 0.05.
#### 2- Therefore, we reject the Null Hypothesis.
#### 3- There is statistically significant evidence of an association between airline and delay category.

### Business Interpretation

#### 1- Delay patterns differ significantly across airlines.
#### 2- Some airlines may experience higher proportions of delayed flights.
#### 3- Airlines should identify the operational factors behind these differences.
#### 4- Airline-specific delay management can improve overall flight performance.

## ➤ANOVA

### Hypotheses

#### 1- Null Hypothesis (H₀):All airlines have the same average arrival delay.
#### 2- Alternate Hypothesis (H₁):At least one airline has a different average arrival delay.
#### 3- Significance Level (α):0.05

### Perform One-Way ANOVA

In [28]:
from scipy.stats import f_oneway

airline_groups = [
    group['ARRIVAL_DELAY'].dropna()
    for _, group in df.groupby('AIRLINE')
]

f_stat, p_value = f_oneway(*airline_groups)

print("F-Statistic:", round(f_stat, 4))
print("P-Value:", round(p_value, 6))

F-Statistic: 20.3707
P-Value: 0.0


### Decision

In [29]:
alpha = 0.05

if p_value < alpha:
    print("Reject the Null Hypothesis")
else:
    print("Fail to Reject the Null Hypothesis")

Reject the Null Hypothesis


#### 1- The p-value (0.0) is less than 0.05.
#### 2- Therefore, we reject the Null Hypothesis.
#### 3- There is statistically significant evidence that average arrival delays differ among airlines.

### Business Interpretation

#### 1- Arrival delay performance differs significantly across airlines.
#### 2- Some airlines have higher average delays than others.
#### 3- Airlines should identify the operational causes of these differences.
#### 4- Airline-specific delay reduction strategies can improve punctuality.

# State:

## ➤Null Hypothesis

#### 1- The Null Hypothesis assumes there is no significant difference between the groups.
#### 2- In our ANOVA, it assumes all airlines have the same average arrival delay.
#### 3- Any observed difference is considered due to random variation.
#### 4- We test this assumption using the ANOVA test.
#### 5- If the p-value is less than 0.05, we reject the Null Hypothesis.

## ➤Alternate Hypothesis

#### 1- The Alternate Hypothesis assumes there is a significant difference between the groups.
#### 2- In our ANOVA, at least one airline has a different average arrival delay.
#### 3- This means airline performance may affect arrival delays.
#### 4- The observed differences are unlikely to be due only to random variation.
#### 5- If the p-value is less than 0.05, we reject the Null Hypothesis in favor of the Alternate Hypothesis.

## ➤Test Statistic


#### 1- The test statistic is the value calculated from the statistical test.
#### 2- It measures how strongly the sample results differ from the Null Hypothesis.
#### 3- In our One-Way ANOVA, the test statistic is called the **F-Statistic**.
#### 4- Our calculated F-Statistic is **20.3707**.
#### 5- A larger F-Statistic indicates greater differences between airline average arrival delays.

## ➤p-value

#### 1- The p-value shows how likely the observed result is if the Null Hypothesis is true.
#### 2- A smaller p-value indicates stronger evidence against the Null Hypothesis.
#### 3- In our ANOVA, the p-value is **0.0**.
#### 4- Since **0.0 < 0.05**, the result is statistically significant.
#### 5- Therefore, we reject the Null Hypothesis.

## ➤Decision

#### 1- The significance level is **0.05**.
#### 2- The p-value is **0.0**.
#### 3- Since **p-value < 0.05**, we reject the Null Hypothesis.
#### 4- There is a significant difference in average arrival delays among airlines.
#### 5- Therefore, airline performance is significantly associated with arrival delay.

## ➤Business Interpretation

#### 1- Average arrival delays differ significantly among airlines.
#### 2- Some airlines may have better on-time performance than others.
#### 3- Airlines should identify the operational causes of higher delays.
#### 4- Airline-specific delay management can improve punctuality.
#### 5- Reducing delays can improve overall operational efficiency and passenger satisfaction.

# 3. Confidence Interval Analysis
### Construct confidence intervals for appropriate numerical features.
### Interpret the results.


In [30]:
from scipy.stats import t

arrival_delay = df['ARRIVAL_DELAY'].dropna()

mean = arrival_delay.mean()
std_error = arrival_delay.sem()
confidence = 0.95

ci = t.interval(
    confidence,
    df=len(arrival_delay) - 1,
    loc=mean,
    scale=std_error
)

print("Mean Arrival Delay:", round(mean, 2))
print("95% Confidence Interval:", 
      round(ci[0], 2), "to", round(ci[1], 2))

Mean Arrival Delay: 4.0
95% Confidence Interval: 3.67 to 4.33


#### 1- The mean arrival delay is **4.00 minutes**.
#### 2- The 95% confidence interval ranges from **3.67 to 4.33 minutes**.
#### 3- We are 95% confident that the true average arrival delay lies within this range.
#### 4- The interval is relatively narrow, indicating a precise estimate.
#### 5- This suggests that the average arrival delay is consistently around 4 minutes.

# 4. Feature Engineering


### Create meaningful features using domain knowledge and insights obtained from previous sprints.
#### Possible examples:

## ➤Delay Category

In [31]:
def delay_category(delay):
    if delay <= 0:
        return 'On Time/Early'
    elif delay <= 15:
        return 'Minor Delay'
    elif delay <= 60:
        return 'Moderate Delay'
    else:
        return 'Severe Delay'

df['DELAY_CATEGORY'] = df['ARRIVAL_DELAY'].apply(delay_category)

In [32]:
df['DELAY_CATEGORY'].value_counts()

DELAY_CATEGORY
On Time/Early     31227
Minor Delay        9083
Moderate Delay     6165
Severe Delay       3525
Name: count, dtype: int64

#### 1- Converts numerical arrival delays into meaningful categories.
#### 2- Makes flight performance easier to understand.
#### 3- Helps compare delay patterns across airlines and airports.
#### 4- Supports identification of flights requiring greater operational attention.

## ➤Delay Severity

In [33]:
def delay_severity(delay):
    if delay <= 0:
        return 0
    elif delay <= 15:
        return 1
    elif delay <= 60:
        return 2
    else:
        return 3

df['DELAY_SEVERITY'] = df['ARRIVAL_DELAY'].apply(delay_severity)

In [34]:
df['DELAY_SEVERITY'].value_counts().sort_index()

DELAY_SEVERITY
0    31227
1     9083
2     6165
3     3525
Name: count, dtype: int64

### Observation
#### 1- Most flights have Severity 0 (31,227 flights).
#### 2- 9,083 flights have Minor Severity.
#### 3- 6,165 flights have Moderate Severity.
#### 4- 3,525 flights have Severe Severity.
#### 5- The feature assigns higher scores to more serious delays.

### Justification

#### 1- Converts delay categories into numerical severity scores.
#### 2- Makes delay severity easier to compare.
#### 3- Helps identify flights with serious operational delays.
#### 4- Higher scores indicate greater delay impact.
#### 5- Useful for future analysis and predictive modeling.

## ➤Flight Distance Category

In [35]:
def distance_category(distance):
    if distance <= 500:
        return 'Short'
    elif distance <= 1500:
        return 'Medium'
    elif distance <= 3000:
        return 'Long'
    else:
        return 'Very Long'

df['DISTANCE_CATEGORY'] = df['DISTANCE'].apply(distance_category)

In [36]:
df['DISTANCE_CATEGORY'].value_counts()

DISTANCE_CATEGORY
Medium       24827
Short        18348
Long          6719
Very Long      106
Name: count, dtype: int64

### Observation
#### 1- Medium-distance flights are the most common with 24,827 flights.
#### 2- Short-distance flights account for 18,348 flights.
#### 3- Long-distance flights account for 6,719 flights.
#### 4- Very Long flights are rare, with only 106 flights.
#### 5- The dataset is mainly composed of short- and medium-distance flights.

### Justification

#### 1- Converts numerical distance into meaningful flight groups.
#### 2- Makes comparison between different flight lengths easier.
#### 3- Helps analyze whether flight distance is related to delays.
#### 4- Supports better scheduling and operational planning.
#### 5- Helps airlines understand the operational requirements of different flight types.

## ➤Peak Hour Indicator

In [37]:
df['SCHEDULED_DEPARTURE'].head()

0    1810
1    1837
2     730
3     622
4     725
Name: SCHEDULED_DEPARTURE, dtype: int64

In [38]:
df['SCHEDULED_HOUR'] = (df['SCHEDULED_DEPARTURE'] // 100).astype(int)

df['SCHEDULED_HOUR'].head()

0    18
1    18
2     7
3     6
4     7
Name: SCHEDULED_HOUR, dtype: int64

In [39]:
df['PEAK_HOUR_INDICATOR'] = df['SCHEDULED_HOUR'].apply(
    lambda x: 1 if (6 <= x < 10) or (17 <= x < 21) else 0
)

In [40]:
df['PEAK_HOUR_INDICATOR'].value_counts()

PEAK_HOUR_INDICATOR
0    25550
1    24450
Name: count, dtype: int64

### Observation

#### 1- 24,450 flights were scheduled during peak hours.
#### 2- 25,550 flights were scheduled during non-peak hours.
#### 3- Non-peak flights are slightly more common.
#### 4- The distribution is relatively balanced between both categories.
#### 5- This feature helps analyze whether peak-hour operations are associated with delays.

### Peak Hour Indicator – Justification

#### 1- Identifies flights scheduled during high-demand operating periods.
#### 2- Helps compare delays between peak and non-peak hours.
#### 3- Supports better crew and aircraft scheduling.
#### 4- Helps airports plan resources during busy periods.
#### 5- Can identify whether peak-hour operations contribute to delays.

### ➤Weekend Flight Indicator

In [41]:
df['WEEKEND_FLIGHT_INDICATOR'] = df['DAY_OF_WEEK'].apply(
    lambda x: 1 if x in [6, 7] else 0
)

In [42]:
df['WEEKEND_FLIGHT_INDICATOR'].value_counts()

WEEKEND_FLIGHT_INDICATOR
0    36779
1    13221
Name: count, dtype: int64

### Weekend Flight Indicator – Observation
#### 1- 36,779 flights were scheduled on weekdays.
#### 2- 13,221 flights were scheduled on weekends.
#### 3- Weekday flights are more common than weekend flights.
#### 4- Weekend flights represent a smaller share of the dataset.
#### 5- This feature helps compare weekday and weekend flight operations.

## ➤Busy Airport Flag

### Count flights per airport

In [43]:
airport_counts = df['ORIGIN_AIRPORT'].value_counts()

airport_counts.head(10)

ORIGIN_AIRPORT
ATL    3002
ORD    2501
DFW    2008
LAX    1722
DEN    1649
PHX    1312
SFO    1262
IAH    1242
LAS    1136
MSP     943
Name: count, dtype: int64

### Define a busy airport

In [44]:
busy_airports = airport_counts[airport_counts > 1000].index

### Create Busy Airport Flag

In [45]:
df['BUSY_AIRPORT_FLAG'] = df['ORIGIN_AIRPORT'].apply(
    lambda x: 1 if x in busy_airports else 0
)

### Check the result

In [46]:
df['BUSY_AIRPORT_FLAG'].value_counts()

BUSY_AIRPORT_FLAG
0    34166
1    15834
Name: count, dtype: int64

### Busy Airport Flag – Justification
#### 1- Identifies airports handling a high number of flights.
#### 2- Helps compare operations between busy and regular airports.
#### 3- Supports analysis of airport congestion and delays.
#### 4- Helps airlines plan resources at high-traffic airports.
#### 5- Can assist in improving airport scheduling and operational efficiency.

## ➤Airline Performance Score

### Calculate average arrival delay by airline

In [47]:
airline_avg_delay = df.groupby('AIRLINE')['ARRIVAL_DELAY'].mean()

airline_avg_delay

AIRLINE
AA     2.253574
AS    -1.172071
B6     5.862357
DL     0.258801
EV     6.514786
F9    10.403509
HA     1.084570
MQ     6.504746
NK    14.507677
OO     5.100737
UA     4.344868
US     2.910349
VX     7.042831
WN     4.385050
Name: ARRIVAL_DELAY, dtype: float64

### Create the Performance Score

In [48]:
airline_performance_score = (
    100 - (airline_avg_delay / airline_avg_delay.max()) * 100
)

airline_performance_score.round(2)

AIRLINE
AA     84.47
AS    108.08
B6     59.59
DL     98.22
EV     55.09
F9     28.29
HA     92.52
MQ     55.16
NK      0.00
OO     64.84
UA     70.05
US     79.94
VX     51.45
WN     69.77
Name: ARRIVAL_DELAY, dtype: float64

### Add the score to the dataset

In [49]:
df['AIRLINE_PERFORMANCE_SCORE'] = df['AIRLINE'].map(
    airline_performance_score
)

### Check the result

In [50]:
df[['AIRLINE', 'AIRLINE_PERFORMANCE_SCORE']].drop_duplicates().sort_values(
    'AIRLINE_PERFORMANCE_SCORE',
    ascending=False
)

,AIRLINE,AIRLINE_PERFORMANCE_SCORE
8,AS,108.078974
4,DL,98.216107
21,HA,92.524167
17,AA,84.466332
0,US,79.939248
3,UA,70.051249
11,WN,69.774278
6,OO,64.841116
26,B6,59.591344
1,MQ,55.163419


### Airline Performance Score – Justification
#### 1- Measures airline performance using average arrival delay.
#### 2- Converts delay performance into an easy-to-understand score.
#### 3- Higher scores represent better punctuality.
#### 4- Helps compare operational performance across airlines.
#### 5- Supports identifying airlines that need greater delay-management attention.

## ➤Delay Risk Category

### Create Delay Risk Category

In [51]:
def delay_risk(delay):
    if delay <= 0:
        return 'Low Risk'
    elif delay <= 15:
        return 'Moderate Risk'
    elif delay <= 60:
        return 'High Risk'
    else:
        return 'Very High Risk'

df['DELAY_RISK_CATEGORY'] = df['ARRIVAL_DELAY'].apply(delay_risk)

### Check the results

In [52]:
df['DELAY_RISK_CATEGORY'].value_counts()

DELAY_RISK_CATEGORY
Low Risk          31227
Moderate Risk      9083
High Risk          6165
Very High Risk     3525
Name: count, dtype: int64

### Delay Risk Category – Observation
#### 1- 31,227 flights are classified as Low Risk.
#### 2- 9,083 flights are classified as Moderate Risk.
#### 3- 6,165 flights are classified as High Risk.
#### 4- 3,525 flights are classified as Very High Risk.
#### 5- Most flights have a low delay risk, while very high-risk flights are less frequent.

# 5. Feature Evaluation
### Evaluate whether the engineered features improve data understanding.
#### Provide supporting evidence.


### Check engineered features

In [53]:
engineered_features = [
    'DELAY_CATEGORY',
    'DELAY_SEVERITY',
    'DISTANCE_CATEGORY',
    'PEAK_HOUR_INDICATOR',
    'WEEKEND_FLIGHT_INDICATOR',

    
    'AIRLINE_PERFORMANCE_SCORE',
    'DELAY_RISK_CATEGORY'
]

df[engineered_features].head()

,DELAY_CATEGORY,DELAY_SEVERITY,DISTANCE_CATEGORY,PEAK_HOUR_INDICATOR,WEEKEND_FLIGHT_INDICATOR,AIRLINE_PERFORMANCE_SCORE,DELAY_RISK_CATEGORY
0,Minor Delay,1,Long,1,0,79.939248,Moderate Risk
1,Moderate Delay,2,Short,1,1,55.163419,High Risk
2,On Time/Early,0,Medium,1,0,79.939248,Low Risk
3,On Time/Early,0,Long,1,0,70.051249,Low Risk
4,On Time/Early,0,Medium,1,0,98.216107,Low Risk


### Feature Evaluation – Supporting Evidence
#### 1- The engineered features are successfully integrated into the dataset.
#### 2- Delay Category and Delay Risk Category clearly classify flight performance.
#### 3- Delay Severity provides a numerical measure of delay impact.
#### 4- Distance, peak-hour, and weekend indicators provide additional operational context.
#### 5- Airline Performance Score allows comparison of airline punctuality.
#### 6- These features make the dataset easier to analyze from a business perspective.

### Supporting evidence

In [54]:
pd.crosstab(
    df['AIRLINE'],
    df['DELAY_CATEGORY']
)

DELAY_CATEGORY,Minor Delay,Moderate Delay,On Time/Early,Severe Delay
AIRLINE,,,,
AA,1115,731,3950,412
AS,279,159,1029,48
B6,395,310,1410,204
DL,1207,665,5256,360
EV,926,644,2939,460
F9,133,123,427,67
HA,200,65,404,9
MQ,370,351,1515,323
NK,224,208,451,107


### Airline Delay Category – Key Findings
#### 1- Most airlines have a higher number of **On Time/Early** flights than delayed flights.
#### 2- **WN** has the highest number of On Time/Early flights (6,714), but also has a high number of Minor and Moderate Delays due to its
#### large flight volume.
#### 3- **DL** shows strong on-time performance with 5,256 On Time/Early flights.
#### 4- **AA** also has a large number of On Time/Early flights (3,950).
#### 5- **EV** has 460 Severe Delay flights, indicating a higher number of serious delays.
#### 6- **F9** and **NK** have relatively high Severe Delay counts compared with their total flights.
#### 7- Airlines should focus on reducing Moderate and Severe Delays to improve overall punctuality.

In [55]:
# compare Distance Category with Arrival Delay:

df.groupby('DISTANCE_CATEGORY')['ARRIVAL_DELAY'].mean().round(2)

DISTANCE_CATEGORY
Long          0.42
Medium        4.30
Short         4.84
Very Long    16.61
Name: ARRIVAL_DELAY, dtype: float64

In [56]:
# Compare Peak Hour with Arrival Delay:

df.groupby('PEAK_HOUR_INDICATOR')['ARRIVAL_DELAY'].mean().round(2)

PEAK_HOUR_INDICATOR
0    4.44
1    3.53
Name: ARRIVAL_DELAY, dtype: float64

In [57]:
# Compare Weekend vs Weekday with Arrival Delay:

df.groupby('WEEKEND_FLIGHT_INDICATOR')['ARRIVAL_DELAY'].mean().round(2)

WEEKEND_FLIGHT_INDICATOR
0    4.53
1    2.51
Name: ARRIVAL_DELAY, dtype: float64

### Feature Evaluation

#### 1- Engineered features convert complex numerical values into meaningful categories and scores.
#### 2- Delay Category and Delay Risk Category make different levels of flight delays easier to understand.
#### 3- Delay Severity provides a numerical representation of delay impact.
#### 4- Distance Category helps compare short, medium, and long-distance flights.
#### 5- Peak Hour and Weekend Indicators help analyze scheduling-related patterns.
#### 6- Airline Performance Score provides an easy comparison of airline punctuality.
#### 7- Group-based comparisons provide supporting evidence for evaluating these features.
#### 8- Overall, the engineered features improve data understanding and support business-oriented analysis.

# 6. Final Business Recommendations


### Summarize the overall findings.
#### Provide practical recommendations that could help airlines:


## ➤Improve on-time performance


#### 1- Monitor airlines with higher average arrival delays and lower performance scores.
#### 2- Focus on reducing Moderate and Severe Delay flights.
#### 3- Improve aircraft turnaround and ground handling efficiency.
#### 4- Provide additional resources during peak operating hours.
#### 5- Use delay-risk categories to identify flights that need early intervention.
#### 6- Regularly track airline-wise delay performance to improve punctuality.

###  -Practical Recommendations

#### 1- Prioritize flights and routes with consistently higher arrival delays.
#### 2- Increase ground staff and resources during peak operating hours.
#### 3- Improve aircraft turnaround procedures to reduce departure delays.
#### 4- Use airline performance scores to identify areas needing improvement.
#### 5- Apply early intervention for flights classified as High or Very High Risk.
#### 6- Review delay patterns regularly and adjust operations accordingly.

## ➤Reduce operational delays


#### 1- Identify the major causes of departure and arrival delays.
#### 2- Focus on flights with Moderate, High, and Severe delay risk.
#### 3- Improve aircraft turnaround and ground-handling processes.
#### 4- Monitor busy airports to reduce congestion-related delays.
#### 5- Allocate additional staff and resources during peak hours.
#### 6- Use delay patterns to take preventive action before delays increase.

### -Practical Recommendations

#### 1- Identify and address the main causes of departure and arrival delays.
#### 2- Improve aircraft turnaround and ground-handling processes.
#### 3- Allocate additional staff and equipment at busy airports.
#### 4- Provide extra operational support during peak hours.
#### 5- Monitor High and Very High Risk flights and take preventive action early.
#### 6- Use historical delay patterns to improve daily operational planning.

## ➤Optimize scheduling

#### 1- Flight delays are influenced by factors such as peak-hour operations, departure delays, and airport workload.
#### 2- Peak-hour flights can create additional pressure on airport resources.
#### 3- Busy airports may require better coordination and scheduling.
#### 4- Flight distance and operational conditions should be considered when planning schedules.
#### 5- Delay-risk information can help airlines identify flights that may require additional planning.

### – Practical Recommendations

#### 1- Avoid scheduling too many flights during peak operating hours.
#### 2- Add sufficient buffer time between tightly scheduled flights.
#### 3- Allocate more aircraft, crew, and ground staff during busy periods.
#### 4- Adjust schedules at airports with consistently high traffic and delays.
#### 5- Use historical delay patterns when planning flight timings.
#### 6- Give additional turnaround time to flights with higher delay risk.

## ➤Improve passenger satisfaction


#### 1- Flight delays can negatively affect the passenger experience.
#### 2- Severe and high-risk delays require greater attention.
#### 3- Better on-time performance can improve customer satisfaction.
#### 4- Clear communication during delays can reduce passenger frustration.
#### 5- Reliable scheduling helps passengers plan their travel with confidence.

### – Practical Recommendations

#### 1- Provide passengers with timely updates about delays and schedule changes.
#### 2- Send alerts through SMS, email, or mobile apps when delays occur.
#### 3- Offer quick rebooking options for significantly delayed or cancelled flights.
#### 4- Provide appropriate assistance to passengers during long delays.
#### 5- Improve punctuality to reduce waiting time and travel disruptions.
#### 6- Use passenger feedback to identify and improve common service issues.

## ➤Reduce cancellation rates

#### 1- Cancellation reasons should be monitored to identify the major causes of cancellations.
#### 2- Weather, airline, and operational issues can affect flight reliability.
#### 3- High cancellation periods should be identified through historical data.
#### 4- Better scheduling and resource planning can help reduce avoidable cancellations.
#### 5- Reducing cancellations improves operational reliability and passenger satisfaction.

### – Practical Recommendations

#### 1- Analyze cancellation reasons regularly and address the most frequent causes.
#### 2- Keep backup aircraft and crew available during high-risk periods.
#### 3- Improve weather monitoring and prepare alternative operating plans.
#### 4- Avoid overloading busy airports with tightly scheduled flights.
#### 5- Provide early rebooking options when cancellation becomes likely.
#### 6- Maintain regular aircraft inspections to reduce maintenance-related cancellations.